test/with_dummyserver/test_socketlevel.py

```diff
@@ -239,6 +239,34 @@ def socket_handler(listener):
         finally:
             timed_out.set()
 
+    def test_delayed_body_read_timeout_with_preload(self):
+        timed_out = Event()
+
+        def socket_handler(listener):
+            sock = listener.accept()[0]
+            buf = b''
+            body = 'Hi'
+            while not buf.endswith(b'\r\n\r\n'):
+                buf += sock.recv(65536)
+            sock.send(('HTTP/1.1 200 OK\r\n'
+                       'Content-Type: text/plain\r\n'
+                       'Content-Length: %d\r\n'
+                       '\r\n' % len(body)).encode('utf-8'))
+
+            timed_out.wait()
```

Comment:
The finally block *should* work, but in general it's a bit better to avoid the risk of deadlocking. It also prevents later changes to this code from missing that subtlety and accidentally removing the finally block.